### ISSUE TITLE
`cargo test --all --lib` doesn't work on virtual manifests

### ISSUE THREAD

**Issue by behnam** [rust-lang/cargo#4387]
The
```
cargo test --all
```
variation works fine in a virtual manifest, but surprisingly
```
cargo test --all --lib
```
doesn't, erring with:
```
error: no library targets found
```

I believe this is a bug, as the filters do make complete sense.

**Comment by alexcrichton** [comment]
Hm yeah I think I'd imagine a better error just saying you can't compose `--all` with more filters?

**Comment by behnam** [comment]
Oh, I'm doing that all the time! Specifically, `test --all --lib` works fine with a non-virtual workspace, it just panics in a virtual workspace.

**Comment by alexcrichton** [comment]
Hm ok, then yeah I guess we just need to implement and/or fix this, and interpret `--all --lib` as "test all libraries"?

**Comment by behnam** [comment]
Yeah, I think falls into the category of not-still-implemented-`--all` feature.

**Comment by rocallahan** [comment]
`cargo check --all --all-targets` fails with the same error in a virtual workspace. I guess that's basically the same as this bug, or should I file a new issue?

**Comment by ehuss** [comment]
I believe this is fixed by #5873.

**Comment by inferiorhumanorgans** [comment]
I'm not 100% sure I'm not doing something wrong, but `cargo test --all --lib` doesn't appear to hit all the libraries defined in a virtual manifest.
